remote_file_system/server.py

```diff
@@ -58,80 +61,122 @@ def listen_for_messages(self) -> None:
 
     def _dispatch_message(self, message: Message, client_ip_address: IPv4Address, client_port_number: int) -> None:
         if isinstance(message, ReadFileRequest):
-            content: bytes = self.server_file_system.read_file(relative_file_path=message.file_name)
-            is_successful, modification_timestamp = self.server_file_system.get_modified_timestamp(message.file_name)
-            reply: ReadFileResponse = ReadFileResponse(
-                reply_id=uuid4(), content=content, modification_timestamp=modification_timestamp
-            )
+            if self._check_for_duplicate_request_message(message):
+                logger.info(f"Duplicate request message detected: {message}")
+                reply = self._get_message_from_history(message.request_id)
+                logger.info(f"Sending message from history: {reply}")
+            else:
+                content: bytes = self.server_file_system.read_file(relative_file_path=message.file_name)
+                is_successful, modification_timestamp = self.server_file_system.get_modified_timestamp(
+                    message.file_name
+                )
+                reply: ReadFileResponse = ReadFileResponse(
+                    reply_id=uuid4(), content=content, modification_timestamp=modification_timestamp
+                )
+                self._add_message_to_history(message.request_id, reply)
```

Comment:
I noticed this code block is duplicated, do you think we could move it to be directly below `def _dispatch_message`? Since the same check for duplication is being performed for all incoming requests
Something like:
```
if message.request_id in self.message_history:
    send_message(self.message_history[message.request_id]
    return
if isinstance(message, ReadFileRequest):
    ...
```

Reply:
sure